src/3rd_party.c

```diff
@@ -0,0 +1,173 @@
+}
+
+static void print_help(void)
+{
+	print("Usage:\n");
+	print("   swupd 3rd-party [OPTION...]\n\n");
+
+	global_print_help();
+
+	print("Options:\n");
+	print("   --add <repo_name>		Add third party repo\n");
+	print("   --list			List third party repo(s)");
+	print("   --remove <repo_name>		Remove third party repo\n");
+	print("\n");
+}
+
+static const struct option prog_opts[] = {
+	{ "repo", required_argument, NULL, 'R' },
+	{ "add", no_argument, 0, CHOOSE_ADD },
+	{ "list", no_argument, 0, CHOOSE_LIST },
+	{ "remove", no_argument, 0, CHOOSE_REMOVE },
+};
+
+static bool parse_opt(int opt, UNUSED_PARAM char *optarg)
+{
+	switch (opt) {
+	case 'R':
+		if (!optarg) {
+			error("Invalid repo: %s\n\n", optarg);
+		}
+		cmdline_option_repo = optarg;
+		return true;
+	case CHOOSE_ADD:
+		cmdline_option_add = optarg_to_bool(optarg);
+		return true;
+	case CHOOSE_REMOVE:
+		cmdline_option_remove = optarg_to_bool(optarg);
+		return true;
+	case CHOOSE_LIST:
+		cmdline_option_list = optarg_to_bool(optarg);
+		return true;
+	default:
+		return false;
+	}
+	return false;
+}
+
+static const struct global_options opts = {
+	prog_opts,
+	sizeof(prog_opts) / sizeof(struct option),
+	parse_opt,
+	print_help,
+};
+
+static bool parse_options(int argc, char **argv)
+{
+	int ind = global_parse_options(argc, argv, &opts);
+	int opt_mask = 0;
+
+	if (ind < 0) {
+		return false;
+	}
+
+	if (!cmdline_option_repo) {
+		if (cmdline_option_add || cmdline_option_remove) {
+			error("The --repo option is a required argument\n\n");
+			return false;
+		}
+	}
+
+	opt_mask = cmdline_option_add + cmdline_option_remove + cmdline_option_list;
+
+	if (opt_mask > 1) {
+		error("Cant use option --add and --remove and --list at the same time\n\n");
+		return false;
+	} else if (opt_mask == 0) {
+		error("Have to choose atleast one of --add, --remove, --list\n\n");
+		return false;
+	}
+
+	if (argc > ind) {
+		error("Unexpected arguments\n\n");
+		return false;
+	}
+
+	return true;
+}
+
+enum swupd_code third_party_main(int UNUSED_PARAM argc, char UNUSED_PARAM **argv)
+{
+	enum swupd_code ret = SWUPD_OK;
+	const int step_in_third_party = 1;
+	char *repo_path = NULL;
+
+	/* there is no need to report in progress for init at this time */
+
+	if (!parse_options(argc, argv)) {
+		print_help();
+		return SWUPD_INVALID_OPTION;
+	}
+	progress_init_steps("third-party", step_in_third_party);
+
+	ret = swupd_init(SWUPD_NO_ROOT);
+	if (ret != SWUPD_OK) {
+		goto finish;
+	}
+
+	get_repo_path(&repo_path, cmdline_option_repo);
+
+	if (cmdline_option_add) {
+		ret = third_party_add(repo_path);
+	} else if (cmdline_option_remove) {
+		ret = third_party_remove(repo_path);
+	} else if (cmdline_option_list) {
+		ret = third_party_list();
+	}
+
+	swupd_deinit();
+
+finish:
+	progress_finish_steps(ret);
+	free_string(&repo_path);
+	return ret;
+}
+
+#else
+
+enum swupd_code third_party_main(int UNUSED_PARAM argc, char UNUSED_PARAM **argv)
+{
+	info("This feature is disabled at compile time\n");
```

Comment:
It's better to avoid a print here so for users this feature won't exist before it's released officially.

Reply:
Makes sense will change in next push